File: src/main/java/com/qtech/bubbles/gui/OptionsButton.java
package com.qtech.bubbles.gui;

import com.qtech.bubbles.QBubbles;
import com.qtech.bubbles.common.gametype.AbstractGameType;
import com.qtech.bubbles.common.interfaces.Listener;
import com.qtech.bubbles.core.controllers.MouseController;
import com.qtech.bubbles.core.utils.categories.GraphicsUtils;
import com.qtech.bubbles.event.MouseEvent;
import com.qtech.bubbles.event.RenderEventPriority;
import com.qtech.bubbles.event.SubscribeEvent;
import com.qtech.bubbles.event.bus.EventBus;
import com.qtech.bubbles.event.type.MouseEventType;
import com.qtech.bubbles.graphics.Border;
import com.qtech.bubbles.graphics.OuterBorder;
import com.qtech.bubbles.media.AudioSlot;
import com.qtech.bubbles.util.Util;
import org.jetbrains.annotations.Nullable;

import java.awt.*;
import java.awt.geom.Rectangle2D;
import java.net.URISyntaxException;
import java.util.Objects;

@SuppressWarnings("unused")
public class OptionsButton extends AbstractButton implements Listener {
    protected long pressedTime;
    protected final long hash;
    protected Rectangle bounds;
    protected RenderEventPriority renderEventPriority;
    protected boolean hovered;
    protected boolean pressed;
    protected boolean eventsActive = false;
    protected Runnable command = () -> {
    };
    protected int clickCount;
    protected String text;
    @Nullable
    protected Integer visualX;
    @Nullable
    protected Integer visualY;
    private EventBus.Handler binding;

    public void setText(String text) {
        this.text = text;
    }

    public String getText() {
        return text;
    }

    public void setVisualX(@Nullable Integer visualX) {
        this.visualX = visualX;
    }

    @Nullable
    public Integer getVisualX() {
        return visualX;
    }


    public void setVisualY(@Nullable Integer visualY) {
        this.visualY = visualY;
    }

    @Nullable
    public Integer getVisualY() {
        return visualY;
    }

    public static class Builder {
        private Rectangle _bounds = new Rectangle(10, 10, 96, 48);
        private String _text = "";
        private RenderEventPriority _renderPriority = RenderEventPriority.AFTER_FILTER;
        private Runnable _command = () -> {
        };

        public Builder() {
        }

        public OptionsButton build() {
            OptionsButton button = new OptionsButton(_bounds.x, _bounds.y, _bounds.width, _bounds.height, _renderPriority);

            button.setText(_text);
            button.setCommand(_command);
            return button;
        }

        public Builder bounds(Rectangle bounds) {
            this._bounds = bounds;
            return this;
        }

        public Builder bounds(int x, int y, int width, int height) {
            this._bounds = new Rectangle(x, y, width, height);
            return this;
        }

        public Builder text(String text) {
            this._text = text;
            return this;
        }

        public Builder renderPriority(RenderEventPriority renderPriority) {
            this._renderPriority = renderPriority;
            return this;
        }

        public Builder command(Runnable command) {
            this._command = command;
            return this;
        }
    }

    protected OptionsButton(int x, int y, int width, int height) {
        this(x, y, width, height, RenderEventPriority.AFTER_FILTER);
    }

    protected OptionsButton(int x, int y, int width, int height, RenderEventPriority renderEventPriority) {
        this.renderEventPriority = renderEventPriority;
        this.bounds = new Rectangle(x, y, width, height);

        hash = System.nanoTime();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        OptionsButton that = (OptionsButton) o;
        return hash == that.hash;
    }

    @Override
    public int hashCode() {
        return Objects.hash(hash);
    }

    @SubscribeEvent
    public void onMouse(MouseEvent evt) {
        if (evt.getType() == MouseEventType.PRESS) {
            if (bounds.contains(evt.getPoint())) {
                pressedTime = System.currentTimeMillis();
                pressed = true;
            }
        } else if (evt.getType() == MouseEventType.RELEASE) {
            if (bounds.contains(evt.getPoint())) {
                command.run();
            }
            pressed = false;
        }
    }

    public void paint(Graphics2D gg) {
        Integer vx = visualX;
        if (visualX == null) vx = getX();
        Integer vy = visualY;
        if (visualY == null) vy = getY();
        Rectangle bounds = new Rectangle(vx, vy, getBounds().width, getBounds().height);

        Point mousePos = MouseController.instance().getCurrentPoint();
        if (mousePos != null) {
            if (bounds.contains(mousePos)) {
                Util.setCursor(QBubbles.getInstance().getPointerCursor());

                if (!hovered) {
                    try {
                        AudioSlot focusChangeSFX = new AudioSlot(Objects.requireNonNull(getClass().getResource("/assets/qbubbles/audio/sfx/ui/button/focus_change.wav")), "focusChange");
                        focusChangeSFX.setVolume(0.2d);
                        focusChangeSFX.play();
                    } catch (URISyntaxException e) {
                        e.printStackTrace();
                    }
                }

                hovered = true;
            } else {
                if (hovered) {
                    Util.setCursor(QBubbles.getInstance().getDefaultCursor());
                    hovered = false;
                }
            }
        }

        Color textColor;

        Stroke oldStroke = gg.getStroke();

        if (pressed && MouseController.instance().getCurrentPoint() != null && getBounds().contains(MouseController.instance().getCurrentPoint())) {

            // Shadow
            Paint old = gg.getPaint();
            GradientPaint p = new GradientPaint(0, vy, new Color(0, 192, 255), 0f, vy + getHeight(), new Color(0, 255, 192));
            gg.setPaint(p);
            gg.fill(bounds);
            gg.setPaint(old);

//            gg.setColor(new Color(0, 96, 128));
//            gg.fill(bounds);
//            gg.setColor(new Color(0, 48, 64));
//            gg.draw(bounds);
            textColor = Color.white;
        } else if (hovered) {

            gg.setStroke(new BasicStroke(2.0f));

            gg.setColor(new Color(128, 128, 128));
            gg.fill(bounds);

            // ShadowQ
//            Paint old = gg.getPaint();
            GradientPaint p = new GradientPaint(0, vy, new Color(0, 192, 255), 0f, vy + getHeight(), new Color(0, 255, 192));
//            gg.setPaint(p);
//            gg.draw(new Rectangle(bounds.x + 1, bounds.y + 1, bounds.width - 2, bounds.height - 2));
//            gg.setPaint(old);
            Border border = new OuterBorder(2, 2, 2, 2);
            border.setPaint(p);
            border.paintBorder(QBubbles.getInstance(), gg, bounds.x, bounds.y, bounds.width, bounds.height);
//            gg.setColor(new Color(0, 192, 192));
//            gg.fill(bounds);
//            gg.setColor(new Color(0, 96, 128));
//            gg.draw(bounds);
            textColor = new Color(255, 255, 255);
        } else {

            gg.setStroke(new BasicStroke(1.0f));

            gg.setColor(new Color(128, 128, 128));
            gg.fill(bounds);

//            gg.setColor(new Color(128, 128, 128));
//            gg.draw(bounds);
            textColor = new Color(192, 192, 192);
        }

        gg.setStroke(oldStroke);

        Graphics2D gg1 = (Graphics2D) gg.create(bounds.x + 1, bounds.y + 1, bounds.width - 2, bounds.height - 2);
        gg1.setColor(textColor);
        GraphicsUtils.drawCenteredString(gg1, text, new Rectangle2D.Double(0, 0, bounds.width - 2, bounds.height - 2), new Font(QBubbles.getInstance().getFont().getName(), Font.BOLD, 16));
        gg1.dispose();
    }

    @Override
    public void destroy() {

    }

    public int getX() {
        return bounds.x;
    }

    public int getY() {
        return bounds.y;
    }

    private int getWidth() {
        return bounds.width;
    }

    public int getHeight() {
        return bounds.height;
    }

    public void tick(AbstractGameType gameType) {

    }

    @Override
    public void bindEvents() {
        QBubbles.getEventBus().register(this);
        eventsActive = true;

        Point mousePos = MouseController.instance().getCurrentPoint();
        boolean hoveredNew = mousePos != null && bounds.contains(mousePos);
        if (mousePos != null && bounds.contains(mousePos)) {
            Util.setCursor(QBubbles.getInstance().getPointerCursor());
            hovered = true;
        }
    }

    @Override
    public void unbindEvents() {
        QBubbles.getEventBus().unregister(this);
        eventsActive = false;

        if (hovered) {
            Util.setCursor(QBubbles.getInstance().getDefaultCursor());
            hovered = false;
        }
    }

    @Override
    public boolean eventsAreActive() {
        return eventsActive;
    }

    public void setBounds(Rectangle bounds) {
        this.bounds = bounds;
    }

    public void setHeight(int height) {
        this.bounds.height = height;
    }

    public void setWidth(int width) {
        this.bounds.width = width;
    }

    public void setX(int x) {
        this.bounds.x = x;
    }

    public void setY(int y) {
        this.bounds.y = y;
    }

    public boolean isHovered() {
        return hovered;
    }

    public boolean isPressed() {
        return pressed;
    }

    public Rectangle getBounds() {
        return bounds;
    }

    public RenderEventPriority getRenderEventPriority() {
        return renderEventPriority;
    }

    public void setRenderEventPriority(RenderEventPriority renderEventPriority) {
        this.renderEventPriority = renderEventPriority;
    }

    public Runnable getCommand() {
        return command;
    }

    public void setCommand(Runnable command) {
        this.command = command;
    }
}

File: src/main/kotlin/com/qtech/bubbles/gui/OptionsButton.kt
package com.qtech.bubbles.gui

import com.qtech.bubbles.BubbleBlaster
import com.qtech.bubbles.common.gametype.AbstractGameMode
import com.qtech.bubbles.common.interfaces.Listener
import com.qtech.bubbles.core.controllers.MouseController.Companion.instance
import com.qtech.bubbles.core.utils.categories.GraphicsUtils.drawCenteredString
import com.qtech.bubbles.event.MouseEvent
import com.qtech.bubbles.event.RenderEventPriority
import com.qtech.bubbles.event.SubscribeEvent
import com.qtech.bubbles.event.type.MouseEventType
import com.qtech.bubbles.graphics.Border
import com.qtech.bubbles.graphics.GraphicsProcessor
import com.qtech.bubbles.media.AudioSlot
import com.qtech.bubbles.util.Util
import java.awt.*
import java.awt.geom.Rectangle2D
import java.net.URISyntaxException
import java.util.*

open class OptionsButton protected constructor(x: Int, y: Int, width: Int, height: Int, var renderEventPriority: RenderEventPriority = RenderEventPriority.AFTER_FILTER) : AbstractButton(), Listener {
    protected var pressedTime: Long = 0
    protected val hash: Long = System.nanoTime()
    var bounds: Rectangle = Rectangle(x, y, width, height)
    var isHovered = false
        protected set
    var isPressed = false
        protected set
    protected var eventsActive = false
    override var command: () -> Unit = { }
    var text: String? = null
    var visualX: Int? = null
    var visualY: Int? = null
    var x: Int
        get() = bounds.x
        set(value) {
            bounds.x = value
        }
    var y: Int
        get() = bounds.y
        set(value) {
            bounds.y = value
        }
    var width: Int
        get() = bounds.width
        set(value) {
            bounds.width = value
        }
    var height: Int
        get() = bounds.height
        set(value){
            bounds.height = value
        }
    var location: Point
        get() = bounds.location
        set(value){
            bounds.location = value
        }
    var size: Dimension
        get() = bounds.size
        set(value){
            bounds.size = value
        }

    class Builder {
        private var bounds = Rectangle(10, 10, 96, 48)
        private var text = ""
        private var renderPriority = RenderEventPriority.AFTER_FILTER
        private var command: () -> Unit = {}
        fun build(): OptionsButton {
            val button = OptionsButton(bounds.x, bounds.y, bounds.width, bounds.height, renderPriority)
            button.text = text
            button.command = command
            return button
        }

        fun bounds(bounds: Rectangle): Builder {
            this.bounds = bounds
            return this
        }

        fun bounds(x: Int, y: Int, width: Int, height: Int): Builder {
            bounds = Rectangle(x, y, width, height)
            return this
        }

        fun text(text: String): Builder {
            this.text = text
            return this
        }

        fun renderPriority(renderPriority: RenderEventPriority): Builder {
            this.renderPriority = renderPriority
            return this
        }

        fun command(command: () -> Unit): Builder {
            this.command = command
            return this
        }
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val that = other as OptionsButton
        return hash == that.hash
    }

    override fun hashCode(): Int {
        return Objects.hash(hash)
    }

    @SubscribeEvent
    fun onMouse(evt: MouseEvent) {
        if (evt.type === MouseEventType.PRESS) {
            if (bounds.contains(evt.point)) {
                pressedTime = System.currentTimeMillis()
                isPressed = true
            }
        } else if (evt.type === MouseEventType.RELEASE) {
            if (bounds.contains(evt.point)) {
                command.invoke()
            }
            isPressed = false
        }
    }

    override fun paint(gp: GraphicsProcessor) {
        var vx = visualX
        if (visualX == null) vx = x
        var vy = visualY
        if (visualY == null) vy = y
        val bounds = Rectangle(vx!!, vy!!, bounds.width, bounds.height)
        val mousePos = instance().currentPoint
        if (mousePos != null) {
            if (bounds.contains(mousePos)) {
                Util.setCursor(BubbleBlaster.instance.pointerCursor)
                if (!isHovered) {
                    try {
                        val focusChangeSFX = AudioSlot(Objects.requireNonNull(javaClass.getResource("/assets/bubbleblaster/audio/sfx/ui/button/focus_change.wav")), "focusChange")
                        focusChangeSFX.volume = 0.2
                        focusChangeSFX.play()
                    } catch (e: URISyntaxException) {
                        e.printStackTrace()
                    }
                }
                isHovered = true
            } else {
                if (isHovered) {
                    Util.setCursor(BubbleBlaster.instance.defaultCursor)
                    isHovered = false
                }
            }
        }
        val textColor: Color
        val oldStroke = gp.stroke()
        if (isPressed && instance().currentPoint != null && bounds.contains(instance().currentPoint!!)) {

            // Shadow
            val old = gp.paint()
            val p = GradientPaint(0F, vy.toFloat(), Color(0, 192, 255), 0f, (vy + height).toFloat(), Color(0, 255, 192))
            gp.paint(p)
            gp.shapeF(bounds)
            gp.paint(old)

//            gg.setColor(new Color(0, 96, 128));
//            gg.fill(bounds);
//            gg.setColor(new Color(0, 48, 64));
//            gg.draw(bounds);
            textColor = Color.white
        } else if (isHovered) {
            gp.stroke(BasicStroke(2.0f))
            gp.color(Color(128, 128, 128))
            gp.shapeF(bounds)

            // ShadowQ
//            Paint old = gg.getPaint();
            val shiftX = bounds.width.toDouble() * 2 * BubbleBlaster.ticks / (BubbleBlaster.TPS * 10)
            val p = GradientPaint(bounds.x + (shiftX.toFloat() - bounds.width), 0F, Color(0, 192, 255), bounds.x + shiftX.toFloat(), 0f, Color(0, 255, 192), true)
            //            gg.setPaint(p);
//            gg.draw(new Rectangle(bounds.x + 1, bounds.y + 1, bounds.width - 2, bounds.height - 2));
//            gg.setPaint(old);
            val border = Border(0, 0, 2, 0)
            border.paint = p
            border.paintBorder(gp, bounds.x, bounds.y, bounds.width, bounds.height)
            //            gg.setColor(new Color(0, 192, 192));
//            gg.fill(bounds);
//            gg.setColor(new Color(0, 96, 128));
//            gg.draw(bounds);
            textColor = Color(255, 255, 255)
        } else {
            gp.stroke(BasicStroke(1.0f))
            gp.color(Color(128, 128, 128))
            gp.shapeF(bounds)

//            gg.setColor(new Color(128, 128, 128));
//            gg.draw(bounds);
            textColor = Color(192, 192, 192)
        }
        gp.stroke(oldStroke)
        val gg1 = gp.create(bounds.x + 1, bounds.y + 1, bounds.width - 2, bounds.height - 2)
        gg1.color(textColor)
        drawCenteredString(gg1, text!!, Rectangle2D.Double(0.0, 0.0, (bounds.width - 2).toDouble(), (bounds.height - 2).toDouble()), Font(BubbleBlaster.instance.font.name, Font.BOLD, 16))
        gg1.dispose()
    }

    override fun destroy() {}
    open fun tick(gameMode: AbstractGameMode?) {}
    override fun bindEvents() {
        BubbleBlaster.eventBus.register(this)
        eventsActive = true
        val mousePos = instance().currentPoint
        if (mousePos != null && bounds.contains(mousePos)) {
            Util.setCursor(BubbleBlaster.instance.pointerCursor)
            isHovered = true
        }
    }

    override fun unbindEvents() {
        BubbleBlaster.eventBus.unregister(this)
        eventsActive = false
        if (isHovered) {
            Util.setCursor(BubbleBlaster.instance.defaultCursor)
            isHovered = false
        }
    }

    override fun eventsAreBound(): Boolean {
        return eventsActive
    }

}